lib/semantics/tools.h

```diff
@@ -248,6 +248,19 @@ template<typename T> std::optional<std::int64_t> GetIntValue(const T &x) {
   }
 }
 
+template<typename T> static bool IsZero(const T &expr) {
+  auto value{GetIntValue(expr)};
+  return value && *value == 0;
+}
+
+template<typename T>
```

Comment:
You could replace the two sites that call this template function with a simple expansion of its body and turn 11 lines into six, and probably improve readability.

Reply:
Thanks, Peter.  I've tightened up the code to emit a message if the expression is zero.

Reply:
I wasn't suggesting that you should remove `IfZero()`.  It's a useful abstraction.  I was suggesting that you remove `SayIfZero()`.

Reply:
Oops.  My misunderstanding.  I've pushed another change that I believe accomplishes what your original comment suggested.